### PR TITLE
Race condition fix: copy BatchPoints into goroutine

### LINKED ISSUE #250: Race Condition in Output Plugins
I deployed a telegraf v0.1.9 instance with multiple output plugins configured. Unfortunately it crashes every few hours. I compiled telegraf with go race detector enabled and it threw following backtrace:
```
==================
WARNING: DATA RACE
Write by goroutine 50:
  runtime.mapassign1()
      /tmp/workdir/go/src/runtime/hashmap.go:411 +0x0
  github.com/influxdb/telegraf/outputs/kafka.(*Kafka).Write()
      /go/src/github.com/influxdb/telegraf/outputs/kafka/kafka.go:68 +0xb76
  github.com/influxdb/telegraf.(*Agent).flush.func1()
      /go/src/github.com/influxdb/telegraf/agent.go:298 +0x10e

Previous write by goroutine 49:
  runtime.mapassign1()
      /tmp/workdir/go/src/runtime/hashmap.go:411 +0x0
  github.com/influxdb/influxdb/client.(*Client).Write()
      /go/src/github.com/influxdb/telegraf/Godeps/_workspace/src/github.com/influxdb/influxdb/client/influxdb.go:191 +0x1088
  github.com/influxdb/telegraf/outputs/influxdb.(*InfluxDB).Write()
      /go/src/github.com/influxdb/telegraf/outputs/influxdb/influxdb.go:128 +0x2f5
  github.com/influxdb/telegraf.(*Agent).flush.func1()
      /go/src/github.com/influxdb/telegraf/agent.go:298 +0x10e

Goroutine 50 (running) created at:
  github.com/influxdb/telegraf.(*Agent).flush()
      /go/src/github.com/influxdb/telegraf/agent.go:302 +0x15a
  github.com/influxdb/telegraf.(*Agent).crankParallel()
      /go/src/github.com/influxdb/telegraf/agent.go:218 +0x706
  github.com/influxdb/telegraf.(*Agent).Run()
      /go/src/github.com/influxdb/telegraf/agent.go:390 +0x218
  main.main()
      /go/src/github.com/influxdb/telegraf/cmd/telegraf/telegraf.go:160 +0x1d4b

Goroutine 49 (running) created at:
  github.com/influxdb/telegraf.(*Agent).flush()
      /go/src/github.com/influxdb/telegraf/agent.go:302 +0x15a
  github.com/influxdb/telegraf.(*Agent).crankParallel()
      /go/src/github.com/influxdb/telegraf/agent.go:218 +0x706
  github.com/influxdb/telegraf.(*Agent).Run()
      /go/src/github.com/influxdb/telegraf/agent.go:390 +0x218
  main.main()
      /go/src/github.com/influxdb/telegraf/cmd/telegraf/telegraf.go:160 +0x1d4b
```
Influxdb and kafka output plugin both modify the BatchPoints datastructure passed to Write func by defining Tags. Those tags are only required locally, so I think copying the Points datastructure before modifying it is in required.

Cya Ed

### PR DESCRIPTION
**Pull request by sparrc**
Fixes #250

### CODE REVIEW

**Comment by sparrc** [comment]
cc @tarrychk

File: agent.go [hunk 1]
```diff
@@ -303,10 +303,18 @@ func (a *Agent) flush(bp *BatchPoints) error {
 
 	for _, o := range a.outputs {
 		wg.Add(1)
+
+		// Copy BatchPoints
+		var bpc BatchPoints
+		bpc.Time = bp.Time
+		bpc.Tags = bp.Tags
+		bpc.Precision = bp.Precision
+		copy(bp.Points, bpc.Points)
```
Comment (jwilder) [inline agent.go]: I think your args might be reversed.  This looks like it will copy the nil `bpc.Points` into `bp.Points`.

File: Makefile [hunk 2]
```diff
@@ -5,7 +5,7 @@ ifndef GOBIN
 endif
 
 build: prepare
-	$(GOBIN)/godep go build -o telegraf -ldflags \
+	$(GOBIN)/godep go build -race -o telegraf -ldflags \
```
Comment (otoolep) [inline Makefile]: You want race-enabled builds permanently?

**Comment by sparrc** [comment]
@jwilder thanks for catching that! this is completely not working as I intended, I need to deep copy the client.Points list

File: Makefile (hunk 2 above)
Comment (sparrc) [inline Makefile]: From the makefile, I think that I do. The packaging script does not use `make` to create it's binaries

Comment (otoolep) [inline Makefile]: OK.

Be aware that enabling race-detection is not free. Your code will run significantly slower.

https://golang.org/doc/articles/race_detector.html#Runtime_Overheads

**Comment by sparrc** [comment]
@jwilder I wrote a deepcopy() function on the BatchPoints struct to deal with the copy

@otoolep I think you're right, probably best to be safe and have a separate `make dev` command in the Makefile

I'm still not 100% sure about this change. It doesn't seem particularly efficient to deepcopy all of the points for each output sink, but I don't see any other way around it if I want the flushing to be multithreaded.

With that being said, I could just have the output sinks flush serially...thoughts?

**Comment by tarrychk** [comment]
Flushing serially wouldn't help if we expect output plugins to modify points. On the other hand if we could rely on output plugins to only read points then passing points by reference would suffice.

On October 8, 2015 10:34:24 PM CEST, Cameron Sparr <notifications@github.com> wrote:
>@jwilder I wrote a deepcopy() function on the BatchPoints struct to
>deal with the copy
>
>@otoolep I think you're right, probably best to be safe and have a
>separate `make dev` command in the Makefile
>
>I'm still not 100% sure about this change. It doesn't seem particularly
>efficient to deepcopy all of the points for each output sink, but I
>don't see any other way around it if I want the flushing to be
>multithreaded.
>
>With that being said, I could just have the output sinks flush
>serially...thoughts?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/influxdb/telegraf/pull/251#issuecomment-146677996

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by sparrc** [comment]
@tarrychk the problem is deeper than that though, the influxdb client itself modifies Tags on the Points struct. https://github.com/influxdb/influxdb/blob/master/client/influxdb.go#L186-L212. (this is probably where the Kafka code came from in the first place)

I'd also just prefer if people writing output plugins didn't need to worry about race conditions within their code.

**Pull request merged**